crates/core/tedge_api/src/entity_store.rs

```diff
@@ -0,0 +1,539 @@
+//! A store containing registered MQTT entities.
+//!
+//! References:
+//!
+//! - <https://github.com/thin-edge/thin-edge.io/issues/2081>
+//! - <https://thin-edge.github.io/thin-edge.io/next/references/mqtt-api/#entity-store>
+
+// TODO: move entity business logic to its own module
+
+use std::collections::HashMap;
+
+use mqtt_channel::Message;
+use mqtt_channel::Topic;
+
+/// Represents an "MQTT entity identifier" portion of the MQTT topic
+///
+/// Example:
+/// - topic: `te/device/dev1/service/myservice/m//my_measurement
+/// - entity id: `device/dev1/service/myservice`
+///
+/// <https://thin-edge.github.io/thin-edge.io/next/references/mqtt-api/#group-identifier>
+type EntityId = String;
+
+// type checker is not able to infer `&EntityId` as &str, so alias is needed
+//
+// TODO: try using Rc, Cow, or string interning to get rid of duplicate strings
+type EntityIdRef<'a> = &'a str;
+
+// In the future, root will be read from config
+const MQTT_ROOT: &str = "te";
+
+/// A store for topic-based entity metadata lookup.
+///
+/// This object is a hashmap from MQTT identifiers to entities (devices or
+/// services) that publish on those topics. It keeps track of type of entities,
+/// their relationships (parent and child entities), and other metadata.
+///
+/// The entity store takes as input registration messages published by entities
+/// (devices and services) and stores information about entities and their
+/// hierarchy, allowing to efficiently query it. It's possible to:
+///
+/// - enumerate all registered devices
+/// - check if a given entity is already registered
+/// - query services and child devices of a given device
+/// - query parent of an entity
+///
+/// # Examples
+///
+/// ```
+/// # use mqtt_channel::{Message, Topic};
+/// # use tedge_api::entity_store::{EntityStore, EntityRegistrationMessage};
+/// let mqtt_message = Message::new(
+///     &Topic::new("te/device/main//").unwrap(),
+///     r#"{"@type": "device"}"#.to_string(),
+/// );
+/// let registration_message = EntityRegistrationMessage::try_from(mqtt_message).unwrap();
+///
+/// let mut entity_store = EntityStore::with_main_device(registration_message);
+/// ```
+#[derive(Debug, Clone)]
+pub struct EntityStore {
+    main_device: EntityId,
+    entities: HashMap<EntityId, EntityMetadata>,
+    external_id_index: HashMap<String, EntityId>,
+}
+
+impl EntityStore {
+    /// Creates a new entity store with a given main device.
+    pub fn with_main_device(main_device: EntityRegistrationMessage) -> Option<Self> {
```

Comment:
This is okay for a first version  but will not work in a setting where the main-device topic is defined by the user using a registration message.

When the mapper starts, the id of the main device is known (This is the CN of its certificate) but not its topic (unless we are using the default topic scheme where the main device is attached to the well-known topic `te/device/main//`).